manifests/params.pp

```diff
@@ -34,7 +34,24 @@
   $postrun_command     = undef
   $dns_alt_names       = []
   $use_srv_records     = false
-  $srv_domain          = $::domain
+
+  # On versions < 3.7.5 there was no defined function as such strict variables cannot be used
+  # On verisons < 4.0.0 and >= 3.5.0 the facts hash was optional disabled by default, on version < 3.5.0 the facts hash did not exist
+  # In some edge cases 'domain' fact may be not defined
+  if versioncmp($::puppetversion, '3.7.5') < 0 {
+    $srv_domain = $::domain
+  } elsif versioncmp($::puppetversion, '4.0.0') >= 0 {
```

Comment:
<pre>
$ puppet --version
4.7.0
$ puppet apply -e "defined('$::domain')"
Notice: Compiled catalog for cobalt in environment production in 0.08 seconds
Notice: Applied catalog in 0.02 seconds
$ puppet apply --modulepath /tmp/mp -e 'pick($::domain, "2nd")'
Warning: Unknown variable: '::domain'.  at line 1:6
Notice: Compiled catalog for cobalt in environment production in 0.06 seconds
Notice: Applied catalog in 0.02 seconds
$ puppet apply --modulepath /tmp/mp -e '$::domain'
Warning: Unknown variable: '::domain'.  at line 1:1
Notice: Compiled catalog for cobalt in environment production in 0.03 seconds
Notice: Applied catalog in 0.03 seconds
</pre>

Only referencing the variable should cause a warning, `defined` should be fine (unless perhaps you're passing the variable into defined rather than the name).